### ISSUE TITLE
Using CPU to create model - MacBook Pro (Retina, 13-inch, Early 2015)

### ISSUE THREAD

**Issue by Walter0807** [apple/turicreate#656]
I'm running the object detection task.
====================================
MacOS Version 10.14 Beta  (18A293u)

MacBook Pro  (Retina, 13-inch, Early 2015) 
Processor 2.7 GHz Intel Core i5 Memory 8 GB 1867 MHz DDR3 
Startup Disk Macintosh HD 
Graphics Intel Iris Graphics 6100 1536 MB

**Comment by srikris** [comment]
Are you using Turi-create 5.0 beta?

**Comment by Walter0807** [comment]
@srikris I've installed turicreate==5.0b1. Should I do something other than that?

**Comment by znation** [comment]
@gustavla Is OD supposed to be GPU accelerated on an `Intel Iris Graphics 6100`? (Perhaps this was tested, and didn't give a significant speedup so it was disabled?)

**Comment by gustavla** [comment]
@Walter0807 Fast GPU acceleration on Mac is only available if it has an AMD GPU (or on older Macs with NVIDIA GPU), but unfortunately not if you only have Intel integrated graphics. Technically, the Metal Performance Shader backend runs on it, but it is not powerful enough to improve over CPU training speeds (so we automatically use that instead).

Let's think about how we can make this clear in either documentation or verbose printing.

**Comment by znation** [comment]
Retagging as docs so we can update the documentation to answer this question.

**Comment by mdenty** [comment]
While performing the training, `turicreate` should display in the log which device is used and eventually the reason of this choice (no mxnet-cu detected, no suitable AMD card found, not implemented yet...) if the training is done on the CPU.

**Comment by b-straub** [comment]
How could you actually test if the build in GPU doesn't speed up anything? When forcing metal usage for object dectection by overwriting "HasHighPowerMetalDevice" I get the following internal metal error:

Using 'image' as feature column
Using 'annotations' as annotations column
Using GPU to create model (Intel Iris Pro Graphics)
+--------------+--------------+--------------+
| Iteration    | Loss         | Elapsed Time |
+--------------+--------------+--------------+
/BuildRoot/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MetalPerformanceShaders-114.8/MPSCore/Utility/MPSLibrary.mm:210: failed assertion `MPSKernel MTLComputePipelineStateCache unable to load function cnnConv_Update_32x32.
        Compiler encountered an internal error: (null)
'
Abort trap: 6

**Comment by srikris** [comment]
@mdenty Great suggestion. That's something we should incorporate.

**Comment by gustavla** [comment]
@b-straub You are right, we do not know this for certain on all platforms. On the ones we have tested, it was slower, but we would love to hear if this is not consistently true, in which case we could revisit the decision mechanism.

I have only seen this be slower though and it has never produced an error like the one you are showing. However, there are two tests inside the `HasHighPowerMetalDevice` functino: 1. Does the device support Metal? 2. Is it a high-power device? So, if you bypass this check and it turns out that the device wasn't just low-power, but it also did not support Metal, failure is expected.

Can you double-check that the device does support Metal? In that's the case, I'll make sure the right bug report is filed to look into this. Thanks for the investigation!

**Comment by b-straub** [comment]
@gustavla Yes the Intel Iris Pro Graphics is supporting Metal just fine. Anyway, I just comment the second part `&& (!dev.isLowPower)` means the device initialization remained untouched.

**Comment by b-straub** [comment]
Should be tagged with "object detection" instead of "image classifier"

**Comment by znation** [comment]
Thanks @b-straub, I'll update the tag.

**Comment by b-straub** [comment]
I tried a fresh build without running multiple make jobs:
- reinstall Xcode 10 beta on Mojave beta
- ensure command line tools are set to Xcode 10
- fresh clone of Turicreate repository

All building well until Metal comes into scope:
`Scanning dependencies of target image_deep_feature_extractor
[ 74%] Building CXX object src/toolkits/image_deep_feature_extractor/CMakeFiles/image_deep_feature_extractor.dir/mlmodel_image_feature_extractor.mm.o
In file included from /Users/xxx/Documents/turicreate/src/toolkits/image_deep_feature_extractor/mlmodel_image_feature_extractor.mm:18:
In file included from /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/CoreML.framework/Headers/CoreML.h:8:
In file included from /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/Foundation.framework/Headers/Foundation.h:6:
In file included from /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CoreFoundation.h:60:
In file included from /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFPropertyList.h:17:
In file included from /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFStream.h:16:
/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CFURL.h:906:91: error: 
      expected ';' after top level declarator
const CFStringRef kCFURLFileProtectionKey API_AVAILABLE(ios(9.0), watchos(2.0), tvos(9.0)) API_PROHIBITED(macos);`
multiple identical errors for missing ";" before API_PROHIBITED

**Comment by gustavla** [comment]
I have forward the Intel Iris Pro Graphics error message to the Metal Performance Shaders developers. As for this error message on the Turi Create side, we don't consider this compute device to be part of what we officially support and the error message does not occur without modifications, so in that regard I think this issue is closed. @b-straub Thanks for the bug report and we do appreciate the scrutiny!

However, we also decided to be more explicit about the compute device decision, so I will leave this issue open until that has been addressed.

**Comment by afranklin** [comment]
Todo: Add runtime message to show which (if any) GPUs are being used and why.

**Comment by cianiandreadev** [comment]
@gustavla can you please confirm that still integrated graphics are not used? Referring to TC version 6.0.
